### DISCUSSION TITLE
[Debian] question about packaging

### DISCUSSION THREAD

**Discussion by werdahias** [nwg-piotr/nwg-shell#18] [Q&A] [answered]
Hi, I want to package your stuff for debian eventually. A few questions regarding that:
- What are the absolute base components for an install? I assume the bar and the panel?
- Do the drawer and the menu depend on the panel or can they be installed standalone?

**Comment by nwg-piotr** [comment]
Hello,

- In principle, each component can be used standalone (well, nwg-panel needs nwg-icon-picker to pick icons, and optionally uses nwg-menu as a plugin).
- As well the drawer, as the menu may be used standalone.
- To have what's called "nwg-shell", you need all the components, together with this repo content (contains `installer` and `check-updates` entry points), and also the [config utility](https://github.com/nwg-piotr/nwg-shell-config).

The main obstacle may be the Debian version. I remember getting in trouble while building the golang stuff (gotk3 library) on some older version. I know little to none on Debian, but will try to be of help, if needed.

**Comment by werdahias** [comment]
All new packages in debian are built in unstable first, that means with the latest dependencies available. What I ment above are the control fields for package [relations](https://www.debian.org/doc/debian-policy/ch-controlfields.html#package-interrelationship-fields-depends-pre-depends-recommends-suggests-breaks-conflicts-provides-replaces-enhances). This means e.g. nwg-panel would `Require` nwg-icons. Can you maybe elaborate a bit on that? E.g. which packages depend on which?
Thanks

**Comment by nwg-piotr** [comment, marked as answer]
Well, [nwg-icon-picker](https://github.com/nwg-piotr/nwg-icon-picker) is a small tool, but it's an nwg-panel dependency. If you click the icon picker icon in the panel settings, it will try to use the `nwg-icon-picker` external command.

![image](https://user-images.githubusercontent.com/20579136/185717382-133442ea-44e0-4f98-b9e8-e062ea2aef1a.png)

If it comes to nwg-menu, if installed, it be may be used by the panel as the MenuStart plugin:

![image](https://user-images.githubusercontent.com/20579136/185717498-2c11b141-8184-4e18-a8ba-8007ee824922.png)

Otherwise the "Menu Start" item will remain inactive.

Dependencies for nwg-panel look like this:

```text
depends=('python' 'python-gobject' 'python-i3ipc' 'python-dasbus' 'python-netifaces' 'python-psutil' 'python-requests' 'bluez-utils' 'gtk3' 'gtk-layer-shell' 'light' 'pamixer' 'playerctl' 'nwg-icon-picker')

makedepends=('python-setuptools' 'python-wheel')

optdepends=('wlr-randr: for non-sway Wayland WMs support'
            'swaync: for Sway Notification Center support'
            'ddcutil: for external displays brightness control')
```

**Comment by werdahias** [comment]
Ok, got it. Seems like I need to package icon picker first, then panel, and then menu. I'll write again once tge packaging is done.

**Comment by nwg-piotr** [comment]
The menu is not really necessary to run the panel, but will be a good test for building gotk3-based golang code.

**Comment by werdahias** [comment]
Quick question about the icon-picker: Is it GPL-licensed? If so, could you maybe add a license file to it?

**Reply by nwg-piotr** [reply]
> Quick question about the icon-picker: Is it GPL-licensed?

It was such a small tool, that I forgot about the license. I decided to release all the nwg-shell components under the same license as sway uses (MIT). File added.

**Comment by werdahias** [comment]
I think I figured out all the dependencies (correct me if I'm wrong):
nwg-shell-config 
-> nwg-menu -> nwg-panel -> nwg-icon-picker
-> nwg-panel
-> nwg-displays
-> nwg-dock -> nwg-drawer
-> nwg-drawer

Standalone stuff:
-> nwg-look
-> nwg-wrapper
-> autotiling

Correct me if I'm wrong.

**Reply by nwg-piotr** [reply]
Please see this: https://aur.archlinux.org/packages/nwg-shell?all_deps=1#pkgdeps

[edit] I'm not sure what to do about baph. It will be useless outside arch and derivatives. We need to think about it, as well as about the panel predefined configs.

**Reply by werdahias** [reply]
Ah thanks, that answers my questions. I feel like this could take a while regardless since some go libraries are missing and sway-nc afaik

**Reply by nwg-piotr** [reply]
I have no idea how, but I managed to build my golang stuff on Debian months ago. I installed Debian (arm) on my ancient Chromebook. It proves it's possible.

**Reply by werdahias** [reply]
yes, building the stuff works fine, I tested it in a debian 11 VM. Debian policy states that every library must be present when building the package (i. e. no downloading), which complicates things.

**Reply by nwg-piotr** [reply]
I spent years on *buntu, but it taught me nothing about Debian. If necessary, I'll try and install Debian on my testing lappy, when I'm back from vacation on Monday.

**Reply by werdahias** [reply]
To be precise: a "local build" works fine, but for a package some libraries are missing.

**Reply by nwg-piotr** [reply]
That's the golang way: include libraries directly from GitHub. Maybe you just need to use binaries?
